data/systemd/snapd.apparmor.service.in

```diff
@@ -12,6 +12,9 @@ Before=sysinit.target
 After=apparmor.service
 ConditionSecurity=apparmor
 RequiresMountsFor=/var/cache/apparmor /var/lib/snapd/apparmor/profiles
+# Rexecution of snapd-apparmor in needs snapd to be mounted
+After=snapd.mounts.target
+Wants=snapd.mounts.target
```

Comment:
doesn't this create a window were the the all snaps are available but the profiles are not loaded yet? how does snap-confine behave in that scenario?

Reply:
There is no dependencies that I have found that makes sure that snapd-apparmor run before mounts. We only ensure it is run before services. Is that not good enough?

If we want to restrict this, we can make a target for only confined mounts. Then make an alias to snapd's mount. Then we add a Before to that confined mounts.

Reply:
Some thing like #13330 would be possible.

Reply:
If we only worry about users manually running applications, we can also make sure that snapd-apparmor is run `Before=systemd-user-sessions.service`.

Reply:
so snap-confine will:

```c
        if (aa_change_onexec(profile) < 0) {
                /* Save errno because secure_getenv() can overwrite it */
                int aa_change_onexec_errno = errno;
                if (secure_getenv("SNAPPY_LAUNCHER_INSIDE_TESTS") == NULL) {
                        errno = aa_change_onexec_errno;
                        if (errno == ENOENT) {
                                fprintf(stderr, "missing profile %s.\n"
                                        "Please make sure that the snapd.apparmor service is enabled and started\n",
                                        profile);
                                exit(1);

```

Reply:
if snapd-apparmor can run before mounts, does that mean it can run before the snapd snap is mounted? and hence it won't re-exec into a newer version of itself as expected in that case (and also this means there is a chance that the system installed apparmor_parser is used rather than the snapd-vendored one and so profile would fail to load)?

Reply:
@alexmurray yes, that's what the bug reported is about

Reply:
I have found out that apparmor.service depends on local-fs.target, and that snapd.apparmor.service depends on apparmor.service, and that all the snaps are before local-fs.target. That means there *must* a window at which the snaps are mounted and snapd-apparmor has not been run. So it has always been like that. I will try to see if I can mount then after local-fs.target. But there will need to several changes.

Reply:
I am not totally sure anymore why apparmor.service is after local-fs.target. I think it is only on some configurations. The upstream service does not have that.

Reply:
*cough* *cough* https://salsa.debian.org/apparmor-team/apparmor/-/blob/ubuntu/master/debian/apparmor.service#L5

Reply:
But this is weird, because there should be re-execution only on those distros.

Reply:
I think the issue was `apparmor.service` being disabled. We should add `Wants=apparmor.service` to `snapd.apparmor.service`.